File: compositor/src/main/java/org/westford/compositor/core/PointerDeviceFactory.java
package org.westford.compositor.core;


import javax.annotation.Nonnull;
import javax.inject.Inject;

public class PointerDeviceFactory {

    @Nonnull
    private final PrivatePointerDeviceFactory privatePointerDeviceFactory;
    @Nonnull
    private final RenderPlatform              renderPlatform;
    @Nonnull
    private final FiniteRegionFactory         finiteRegionFactory;

    @Inject
    PointerDeviceFactory(@Nonnull final PrivatePointerDeviceFactory privatePointerDeviceFactory,
                         @Nonnull final RenderPlatform renderPlatform,
                         @Nonnull final FiniteRegionFactory finiteRegionFactory) {
        this.privatePointerDeviceFactory = privatePointerDeviceFactory;
        this.renderPlatform = renderPlatform;
        this.finiteRegionFactory = finiteRegionFactory;
    }

    public PointerDevice create() {
        final FiniteRegion  outputsRegion = this.finiteRegionFactory.create();
        final PointerDevice pointerDevice = this.privatePointerDeviceFactory.create(outputsRegion);

        this.renderPlatform.getWlOutputs()
                           .forEach(wlOutput ->
                                            outputsRegion.add(wlOutput.getOutput()
                                                                      .getRegion()));
        return pointerDevice;
    }
}

File: compositor/src/main/kotlin/org/westford/compositor/core/PointerDeviceFactory.kt
package org.westford.compositor.core

import javax.inject.Inject

class PointerDeviceFactory @Inject internal constructor(private val privatePointerDeviceFactory: PrivatePointerDeviceFactory,
                                                        private val renderPlatform: RenderPlatform,
                                                        private val finiteRegionFactory: FiniteRegionFactory) {

    fun create(): PointerDevice {
        var outputsRegion = this.finiteRegionFactory.create()

        this.renderPlatform.wlOutputs.forEach {
            outputsRegion += it.output.region
        }

        return this.privatePointerDeviceFactory.create(outputsRegion)
    }
}
